### PR TITLE
Add Promise-based API

### PR DESCRIPTION
**Pull request by brendanashworth**
Issue: https://github.com/minio/minio-js/issues/568

This contains the basis for the new Promise-based API. Right now it makes callbacks *optional*, and when no callback is given, it will instead return a Promise. Backwards compatibility is maintained. Tests and documentation have been added.

Right now this only changes actions dealing with buckets, though it is easily scale-able once we decide this is the way to go. **Seeking feedback!**

### CODE REVIEW

File: src/test/unit/test.js [hunk 1]
```diff
@@ -384,25 +384,37 @@ describe('Client', function() {
       })
```
Comment (krishnasrinivas) [inline src/test/unit/test.js]: you can remove the changes to the `test.js` we are going to remove all the mock tests in test.js

Unit tests will test only functions which do not interact with external world (i.e which make calls to S3)

Comment (brendanashworth) [inline src/test/unit/test.js]: okay! thanks for the heads up

Comment (brendanashworth) [inline src/test/unit/test.js]: fixed!

**Comment by brendanashworth** [comment]
@krishnasrinivas how does this look? Should I convert the rest of the API?

File: src/main/helpers.js [hunk 2]
```diff
@@ -17,6 +17,38 @@
...
+
+    // If the callback is given, don't promisify, just pass straight in.
+    if (callback) return fn.apply(this, arguments)
+
+    // Otherwise, create a new set of arguments, and wrap
+    // it in a promise.
+    let args = []
+    for (let i = 0; i < arguments.length; i++) {
+      args.push(arguments[i])
+    }
```
Comment (krishnasrinivas) [inline src/main/helpers.js]: you can do this: let args = [...arguments]

File: src/main/helpers.js [hunk 3]
```diff
@@ -17,6 +17,38 @@
...
+// Returns a wrapper function that will promisify a given callback function.
+// It will preserve 'this'.
+export function promisify(fn) {
+  return function() {
+    // If the last argument is a function, assume its the callback.
+    let callback = arguments[arguments.length - 1]
+    if (typeof callback !== 'function') callback = null
+
+    // If the callback is given, don't promisify, just pass straight in.
+    if (callback) return fn.apply(this, arguments)
```
Comment (krishnasrinivas) [inline src/main/helpers.js]: more concise:
if (typeof callback === 'function') return fn.apply(this, arguments)

File: src/main/minio.js [hunk 4]
```diff
@@ -2013,6 +2013,12 @@ export class Client {
   }
 }
 
+// Promisify various public-facing APIs on the Client module.
+Client.prototype.makeBucket = promisify(Client.prototype.makeBucket)
+Client.prototype.listBuckets = promisify(Client.prototype.listBuckets)
+Client.prototype.bucketExists = promisify(Client.prototype.bucketExists)
+Client.prototype.removeBucket = promisify(Client.prototype.removeBucket)
```
Comment (krishnasrinivas) [inline src/main/minio.js]: note that we are using es6 classes

Comment (krishnasrinivas) [inline src/main/minio.js]: will this work if we use es6 classes?

Comment (brendanashworth) [inline src/main/minio.js]: Yes, it works right now (with transpiling) and re-assigning class functions works in the REPL (native), if we ever stop transpiling.

File: src/main/minio.js [hunk 5]
```diff
@@ -2013,6 +2013,32 @@ export class Client {
   }
 }
 
+// Promisify various public-facing APIs on the Client module.
+Client.prototype.makeBucket = promisify(Client.prototype.makeBucket)
```
Comment (brendanashworth) [inline src/main/minio.js]: If we'd like something less verbose, I can list the functions as strings and turn this into a `for` loop.

**Comment by brendanashworth** [comment]
I've added the rest of the functions, please review.

File: src/test/functional/functional-tests.js [hunk 6]
```diff
@@ -140,25 +140,38 @@ describe('functional tests', function() {
...
       })
     })
+    it('should list buckets as promise', done => {
+      client.listBuckets()
+        .then(buckets => {
+          if (!_.find(buckets, { name: bucketName })) done(new Error('bucket not found'))
+        })
+        .then(done)
+        .catch(assert.fail)
+    })
```
Comment (krishnasrinivas) [inline src/test/functional/functional-tests.js]:
```
client.listBuckets()
.then(buckets => {
    if (!_.find(buckets, { name: bucketName })) done(new Error('bucket not found'))
    done()
})
.catch(done)
```

File: src/test/functional/functional-tests.js [hunk 7]
```diff
@@ -140,25 +140,39 @@ describe('functional tests', function() {
...
-  describe('makeBucket with region', () => {
     it('should succeed', done => {
       clientUsEastRegion.makeBucket(`${bucketName}.region`, 'us-east-1', done)
     })
     it('should delete bucket', done => {
       clientUsEastRegion.removeBucket(`${bucketName}.region`, done)
     })
+    it('should succeed as promise', done => {
+      clientUsEastRegion.makeBucket(`${bucketName}.region`, 'us-east-1')
+        .then(() => done())
```
Comment (krishnasrinivas) [inline src/test/functional/functional-tests.js]: `.catch(done)` is missing here and at other places as well

**Comment by brendanashworth** [comment]
Thanks!

**Comment by jpospychala** [comment]
hi, are you planning to release this promises support as a new version to npm?
I just stumbled upon your project today and find it super sweet coincidence that you just did, what I need.

**Comment by harshavardhana** [comment]
We will soon @jpospychala

**Pull request merged**